rt/engine/src/main/kotlin/RuntimeEngine.kt

```diff
@@ -38,33 +39,50 @@ import com.yandex.yatagan.validation.impl.GraphValidationExtension
 import com.yandex.yatagan.validation.impl.validate
 import com.yandex.yatagan.validation.spi.ValidationPluginProvider
 import java.lang.reflect.Proxy
+import java.util.concurrent.Future
 import kotlin.system.measureTimeMillis
 
 /**
  * Main entrypoint class for reflection backend implementation.
  */
-class RuntimeEngine<P : RuntimeEngine.Params>(
-    val params: P,
+class RuntimeEngine(
+    private val paramsFuture: Future<Params>,
 ) {
+    private val params: Params
+        get() = paramsFuture.get()
+
+    private val logger: Logger?
+        get() = paramsFuture.get().validationDelegate.logger
+
+    data class Params(
+        var validationDelegate: DynamicValidationDelegate = SimpleDynamicValidationDelegate(),
+        var maxIssueEncounterPaths: Int = 5,
+        var isStrictMode: Boolean = true,
+        var usePlainOutput: Boolean = false,
+    )
+
     init {
-        @OptIn(InternalLangApi::class)
-        LangModelFactory.delegate = RtModelFactoryImpl(javaClass.classLoader)
+        initIfNeeded()
     }
 
-    interface Params {
-        val validationDelegate: DynamicValidationDelegate?
-        val maxIssueEncounterPaths: Int
-        val isStrictMode: Boolean
-        val logger: Logger?
+    private fun initIfNeeded() {
+        @OptIn(InternalLangApi::class)
+        with(LangModelFactory) {
+            if (delegate.get() == null) {
```

Comment:
unnecessary check, compareAndSet will do it

Reply:
Yes, but we don't want to create new factory instance to attempt setting on _every_ call. Hence the pre-check.